### DISCUSSION TITLE
Limit the amount of collision of temporary files on NFS?

### DISCUSSION THREAD

**Discussion by jrosdahl** [ccache/ccache#512] [General]
_From @Romain-Geissler-1A in https://github.com/ccache/ccache/issues/480#issuecomment-564929385_:

It happens that recently in my company we started to hit this issue more and more. We think it is linked to the fact that more and more of our builds are now targetting a single common ccache folder, which is mounted through NFS between hundreads and hundreads of machines. As it's written many times in the code, on NFS file operations aren't as atomic as we could hope for a local filesystem, and most likely problems are coming down from here. Note that we are using more and more docker containers for that, with containers having fixed hostname rather than generated "random" ones, which decreases entropy of the names you generate.

In the end what we did is that we saw your commit, we upgrade ccache, and now are build stop failing "randomly" (still it happened very often). That's cool, but it is just a workaround, in the end the goal of ccache is to avoid rebuilding as much as possible, and here we have cases where we know we rebuild just because we use NFS where atomicity of operations is not guaranteed, not because there is an actual code change. And the more machines we have, the more collision there are, and the more useless rebuilds there are.

So I propose we limit the amount of collision for the NFS case by adding more entropy in the temporary files names. In legacy_util.cpp, function tmp_string you use both hostname and pid to add entropy. In the context of recent linux with container support, using just the hostname and the pid is not enough to try to be "unique". For the hostname I would prepend it with the Linux UTS namespace id. For the PID, I would prepend it with the Linux PID namespace id. How to access namespace ids from C/C++ ? Just call "stat("/proc/self/ns/uts")" and "stat(/proc/self/ns/pid")" if these files do exists, and there the namespace id is actually the inode number "st_ino". Old Linux will not have these files, so you shall cope with it. Make sure you use "stat" and not "lstat", these are symlinks and we want the inode of the file pointed too. Doing that would reduce collisions between ccache run from containers using the same hostname (as we do).

It happens that I also had these issues sometimes one the same machine using the same container, so namespace were the same. I guess this may come from big project spawning a huge amount of very small processes very quickly, and somehow reaching pid_max quickly and reverting back to small pids. So what I would do as well is maybe add more entropy by not only using the pid of the current process, but also the parent pid getppid().

What do you think about the collision reduction mechanisms ?

Cheers,
Romain

**Comment by jrosdahl** [comment]
@Romain-geissler-1A wrote:

> So I propose we limit the amount of collision for the NFS case by adding more entropy in the temporary files names. In legacy_util.cpp, function tmp_string you use both hostname and pid to add entropy. In the context of recent linux with container support, using just the hostname and the pid is not enough to try to be "unique".

Right, using the hostname and PID is not enough to be unique in itself, but that's not the only information that is added to temporary file name: a six bytes string of random numbers and letters is added by mkstemp when creating the file (in addition to making sure that the file is unique using `open` with `O_EXCL`, which might not work on older NFS versions or other strange file systems). On my system mkstemp takes bytes from the set [0-9A-Za-z], so there are 62^6 = 56800235584 combinations to choose from. If you're seeing more than the very occasional collision (on a file system where `O_EXCL` isn't working, that is) then it might be that the random generator seed gets the same. It looks like it's common for libcs to seed the algorithm by the PID and the current microsecond from some system clock.

That said, I'm not against adding UTS and PID namespace IDs to the temporary names, but before doing that I would like to be sure that it actually will help and that there isn't some other kind of problem lurking.

So:

> we have cases where we know we rebuild just because we use NFS where atomicity of operations is not guaranteed, not because there is an actual code change. And the more machines we have, the more collision there are, and the more useless rebuilds there are.

Could you perhaps describe in more detail what you're seeing? How have you ruled out that the problem isn't something else?

Since you seem knowledgeable about both ccache and namespacing, perhaps you could make the needed changes locally and try if they make a difference in your scenario? That would be very useful information. Thanks.

**Comment by Romain-Geissler-1A** [comment]
Well we did hit the issue #480 one day (and honestly I have on idea why it suddenly started from on day to the other), then I decided to migrate to a version of ccache having the fix for #480 (this one: https://github.com/llunak/ccache/commit/22d3a0de52fe4d6fe983cf1e3544fa90ac66e106) and for older versions of ccache which we couldn't update, I just backported it.

Now the builds in my company are not facing this issue, still for me it's just a workaround, not really a fix of the underlying original issue. Since this was an issue happening randomly and requiring many parallel builds of many teams in my company, I can't really investigate an occurence a simple way, and can't make our whole company go back in time and use the unpatched version of ccache just to investigate this problem.

The comment I did was after reviewing your code. In your code you do try to add hostname and pid in the temporary name you generate, most likely to add some "randomness", and I just mentionned that in the scope of containers all using the same hsotname, this might not be so random after all. So feel free to implement my comments as is, other yes, let's just close this.

**Comment by jrosdahl** [comment]
> then I decided to migrate to a version of ccache having the fix for #480 (this one: llunak@22d3a0d) and for older versions of ccache which we couldn't update, I just backported it.

OK, I see. Yes, 22d3a0d is a workaround but e1fc59223518c53ab227d224a36fe2eeead127eb is a real fix: it makes ccache not remove temporary files for other ccache processes when the system clock is changed (unless by several days). (Adding more randomness doesn't fix the issue if  e1fc59223518c53ab227d224a36fe2eeead127eb is not included.)
